### ISSUE TITLE
implement edgecov attributes in tergmLite

### ISSUE THREAD

**Issue by ThomasKraft** [statnet/tergmLite#47]
It would be very helpful if tergmLite's functionality could be expanded to support models that include `edgecov` terms. I have now encountered two scenarios where tergmLite functionality is desired (complex models requiring large MCMC burnin and sample size that cause slow simulation time) but cannot be used because of the necessary inclusion of `edgecov` terms in models .

This linked post (EpiModel example: https://github.com/statnet/EpiModel/issues/495#issuecomment-762832195) includes self-contained code to test the inclusion of an `edgecov` term with tergmLite. Because the tergm package is already built to handle `edgecov` terms, it seems like the only barrier to implementing them in tergmLite is storing them in an accessible format that can be accessed by tergm without storage on a network object. One possible simplistic solution might be to query the network object for edgecov parameters stored in %n%, and store them as a list of matrices in `dat`.

**Comment by smjenness** [comment]
@ThomasKraft : thanks for this suggestion. Our team has discussed and we will consider it for future work potentially later this year. However, we would like to suggest that there may be few if any efficiency gains with this (compared to use full tergm simulation of network objects with edgecov attributes). That is because the edgecov attributes would be stored with an adjacency matrix still in tergmLite. For a network of any substantial size as to require tergmLite (e.g., there are fewer efficiency gains below networks of size 1000), this matrix would require moving around a lot of data, which would likely offset any efficiency gains by representing the core network (nodal attributes and edge list) in a sparse format. Just curious -- what are the network sizes you are working with in your applied models? Thanks!

**Comment by ThomasKraft** [comment]
Thanks for this insight! This makes sense in terms of what I have been experiencing. In my applied models I have ~7500 nodes, and my model includes 3 edgecov's in addition to block constraints and a number of vertex attributes. I also am dealing with short dissolution durations, as well as somewhat computationally expensive custom modules that dynamically alter edgecov variables throughout the simulation.

As you might expect, these factors combine to produce extremely long runtimes for my simulations across a modest number of steps. Alot of this is due to having 3 separate 7500x7500 matrices for my edgecov's, which **have** to be stored as network covariates using %n% in order to run in parallel and modify them dynamically during the simulations. The short dissolution times further require extremely high MCMC burn-in and sample sizes.

An additional problem that arises is memory usage. I'm currently running these models on a HPC, and because I can only use cores on a single node (I believe there used to be netsim_parallel which worked with MPI, but I don't think I can utilize MPI with netsim now), the memory usage becomes extraordinary when I run many cores simultaneously. For example, my `netest` fit object alone is ~16 GB, and I managed to crash a 1.2 Tb cluster running 40 simulations simultaneously on 40 cores. 150 iterations of a single simulation takes ~4 hrs to complete on a fast desktop. 

Unfortunately I don't see clear ways to improve efficiency because the limiting steps are: 1) size of netest fit with network (edge) covariates, 2) simulation time from the TERGM. Any suggestions or ideas would be much appreciated. I had considered using sparse matrix representation but it's important to remember that these are covariates and not network edges, so in my case there are actually very few zeros represented.

**Comment by martinamorris** [comment]
Just FYI, in the next release of the statnet tools, we have implemented a set of new MCMC proposals that dramatically reduce estimation and simulation times.  We are in the testing phases now, and should have a beta release (on GitHub) hopefully in the next couple of weeks.  Would you be interested in trying  those?

I don't think this will solve the edgecov memory issues, but it might reduce the burden of MCMC sampling.

**Comment by ThomasKraft** [comment]
@martinamorris : That's great news! I will try this on my local machine, and since I have access to the development version already I should be able to go ahead (e.g. using remotes::install_github("statnet/ergm-private@dev") ). 

Thus far I've avoided using anything but CRAN versions on the HPC because I've had problems with package/library management. But that should be surmountable.

**Comment by martinamorris** [comment]
@ThomasKraft it's a bit more complicated -- the changes in ergm 4.0 are substantial, and affect many of the packages, so they all need to be updated as a suite.  That's also true for EpiModel -- you'll need the development version that is compatible with the 4.0 tools.

On the statnet side, if you install from ergm-private@dev, it should automatically pull the compatible dependencies.  But I'll need to ask about which branch/commit of EpiModel you'll need.

Before you start in on this, let me also check with @chad-klumb and @krivit to see if the dev branch is in a position to be used right now.  There's lots of working going on to finalize things, so it might be best to wait until we merge the dev branch in to master.

**Comment by ThomasKraft** [comment]
I'll plan on waiting for the dev branch to be merged to master in that case. And when I update to the new master, I'll plan on doing so for both statnet and EpiModel. Thanks as always for the help.

**Comment by smjenness** [comment]
tergmLite has been disbanded (its functionality has been folded into EpiModel itself). If there are any relevant open sub-issues here, please open new issues on the EpiModel or statnet package repositories. Thanks!
